elftoc/elftoc.1
.TH ELFTOC 1 "May 2011" "ELF kickers 3.0"
.LO 1
.SH NAME
elftoc \- translate an ELF file into a C struct
.SH SYNOPSIS
.B elftoc
[OPTIONS] FILE
.SH DESCRIPTION
.B elftoc
takes as input an ELF file, and translates its content into a C source
file describing a struct with an initialized value that matches the
ELF file's memory image. Where possible, literal values are replaced
with preprocessor symbols and sizeof-based calculations.
.SH OPTIONS
.TP
.BR \-e ", " \--force\-end
Force inclusion of a final _end field in the C struct. By default, an
_end field is only included if the padding causes the struct to be
larger than the original ELF file. With this option, a struct with no
extra padding will provide an _end field defined as a zero-length
array.
.TP
.BR \-e ", " \--exclude\-end
Do not include a final _end field, even if padding is present at the
end of the C struct.
.TP
\fB\-w\fR, \fB\--width=\fR\fIN\fR
Set the maximum line width in the output to
.IR N .
The default maximum line width is 80.
.TP
\fB\-i\fR, \fB\--indent=\fR\fIN\fR
Set the number of spaces to indent to
.IR N .
The default indentation amount is 2.
.TP
\fB\-s\fR, \fB\--structname=\fR\fINAME\fR
Set the name of the C struct in the generated source code to
.IR NAME .
By default the struct's name is "elf".
.TP
\fB\-v\fR, \fB\--varname=\fR\fINAME\fR
Set the name of the variable in the generated source code to
.IR NAME .
By default the variable's name is "foo".
.TP
.BR \-m ", " \--allow\-misaligned
A misaligned section in the ELF file would cause padding bytes to be
inserted in the C struct. To avoid this,
.B elftoc
represents misaligned sections as uninterpreted byte arrays,
regardless of their original type. This option suppresses this
behavior, so that misaligned sections retain their actual type, even
when this creates an incorrect image in the C struct.
.TP
.BR \-q ", " \--quiet
Suppress warning messages regarding invalid and contradictory values
in the ELF file.
.TP
\fB\-o\fR, \fB\--output=\fR\fIFILENAME\fR
Write the C source code to
.I FILENAME
instead of to standard output.
.TP
.B \--help
Display help and exit.
.TP
.B \--version
Display version information and exit.
.SH COPYRIGHT
Copyright \(co 1999, 2001, 2011 Brian Raiter
.IR <breadbox@muppetlabs.com> .
.P
License GPLv2+: GNU GPL version 2 or later. This is free software: you
are free to change and redistribute it. There is NO WARRANTY, to the
extent permitted by law.


Changelog
* November 2011: version 3.0a

	This is a very minor update.

	Sofian Brabez supplied me with information to help ELF Kickers
	compile under FreeBSD. I've gratefully incorporated his changes.

	Raphaël Huck reported that sstrip's alteration to the e_shentsize
	field is flagged as invalid by file(1). This has been fixed.

	infect now attempts to reset the infected file's mtime. Not
	important, but why not.
	
* May 2011: version 3.0

	After a decade hiatus, I finally took the plunge and updated this
	package. With some effort I've managed to get a new version out in
	time for the ten-year anniversary of the last update.

	The most important change: All the programs (except ebfc) have
	been updated to deal equally well with 32-bit and 64-bit ELF
	files. This has been a long time coming (though it had to wait
	until I had a 64-bit machine of my own to test on).

	A very simple library, elfrw, has been added which hides many of
	the differences between 32-bit and 64-bit ELF files, allowing the
	smaller programs to support both types of files without
	significant code changes.

	The program objres has been added to the collection.

	The sample program infect has been added to the collection.

	All of the old programs have been cleaned up to avoid the new gcc
	warnings, use getopt_long, and in general brought more in line
	with my current programming style.

	elftoc in particular has had several sections rewritten. The most
	important change is that the build process reads the local copy of
	elf.h to find out which ELF macros are available on the current
	installation, instead of relying on a hard-coded list included in
	the distribution. elftoc also required lots of nearly-duplicate
	code in order to support both 32-bit and 64-bit ELF files.

	Man pages have been added for the utilities that didn't have them,
	for the sake of consistency if nothing else.

	The tiny programs have been removed from ELF kickers. Not because
	I don't love them any less, but they tend to suffer more quickly
	than the rest of the code from changes to Linux. In the last ten
	years I've had to modify nearly all of them at least once to keep
	them working with current kernels.

	(ebfc is the only utility that has not been updated for 64-bit. I
	intend to remedy this in the near future.)
	
* May 2001: version 2.0a

	Two small bug fixes, reported to me by people who downloaded
	version 2.0, the details of which are lost to history.
	
* April 2001: version 2.0

	I began spelunking in the system header files <elf.h> and
	<linux/elf.h> again, and discovered a large number of new
	definitions. I realized that the ELF standard could be going
	through major changes, and that having a version number is really
	necessary.

	elftoc updated to understand some of these new additions, and
	elsewhere restructured so as to make it easier to update it again
	once I understand the rest of them.

	Fixes for elftoc bugs that I noticed during the rewrite.

	elfls updated similarly to elftoc.

	Minor rewrite of sstrip to make the code more airtight and
	aesthetically pleasing.
	
* March 2001

	rebind added to ELF Kickers.

	date and snake added to the tiny collection.

	cat removed from the tiny collection (too uninteresting).

	Cleaned up all tiny programs and made them more stylistically
	consistent.

	Minor bug fix in elftoc.

	Documentation cleanups.

* February 2001

	Five of the tiny programs did not work under 2.2.17 and later 2.2
	kernels due to a change in how Linux checks the ELF header.
	(Thanks to Francois-Rene Rideau, who pointed this out to me. Oddly
	enough, this new restriction does not seem to have made it into
	the 2.4.0 kernel. I am unsure if the change is confined to the 2.2
	versions, or if it will resurface in a later kernel.) I succeeded
	in fixing two of the programs without sacrificing compatibility
	with 2.0 kernels or increasing the size. For the other three, I
	added new versions specifically for 2.2.17 (et al).

* January 2001

	keepalive added to the tiny collection.

	Removed -W option from the Makefile for ebfc (I thought I had done
	this already).

	New feature added to ebfc -- namely, the acceptance of compressed
	source files -- just for the heck of it.

	This file added.

* August 2000
	
	After installing a 2.2 kernel, I discovered that changes to the
	system header files and a new warning in gcc 2.95.2 caused several
	of the programs to generate numerous compiler warnings and errors.
	The errors were caused by a benign redefinition of a macro, and
	were suppressed by changing the inclusion order of the system
	header files in question. The warnings were due to my taking
	advantage of of the standard feature of omitting trailing
	initializers in the definition of a structure, and were suppressed
	by omitting the -W gcc option in the Makefiles, leaving just
	-Wall.

	ls added to the tiny collection.

	Bug fixes and minor improvements to the documentation.

* July 1999
	
	Initial release. Main programs: elfls, elftoc, sstrip, ebfc. In
	the tiny collection: true/false, hello, fgconsole, hexdump, cat,
	factor.


elftoc/elftoc.c
/* elftoc.c: The central module.
 *
 * Copyright (C) 1999-2001 by Brian Raiter, under the GNU General
 * Public License. No warranty. See COPYING for details.
 */

#include	<stdio.h>
#include	<stdlib.h>
#include	<string.h>
#include	<ctype.h>
#include	<stdarg.h>
#include	<errno.h>
#include	<sys/stat.h>
#include	"gen.h"
#include	"elf.h"
#include	"pieces.h"
#include	"addr.h"
#include	"shdrtab.h"
#include	"dynamic.h"
#include	"out.h"
#include	"elftoc.h"

/* The name to use for the initialized variable in the generated C code.
 */
char const	       *varname;

/* The name to use for the structure defined in the generated C code.
 */
char const	       *structname;

/* The output file.
 */
FILE		       *outfile;

/* The name of this program, taken from argv[0]. Used in error
 * messages.
 */
static char const      *thisprog;

/* The name of the input file, taken from argv[1].
 */
static char const      *thefilename = NULL;

/* The input file.
 */
static FILE	       *infile = NULL;

/* The input file's ELF header.
 */
static Elf32_Ehdr      *ehdr = NULL;

/* The input file's program header table, if such is present.
 */
static Elf32_Phdr      *phdrs = NULL;

/* The input file's section header table, if such is present.
 */
static Elf32_Shdr      *shdrs = NULL;

/* The size of the input file.
 */
static int		thefilesize;

/* Displays a formatted error message on stderr. If fmt is NULL, then
 * prints a canned out-of-memory error and exits the program directly.
 * Otherwise, FALSE is returned.
 */
int err(char const *fmt, ...)
{
    va_list	args;

    if (!fmt) {
	fputs("Out of memory!\n", stderr);
	exit(EXIT_FAILURE);
    }
    fprintf(stderr, "%s: ", thefilename ? thefilename : thisprog);
    va_start(args, fmt);
    vfprintf(stderr, fmt, args);
    fputc('\n', stderr);
    va_end(args);
    return FALSE;
}

/* A macro to simplify using errno in reporting errors.
 */
#define	errnomsg(msg)	\
    (err(errno ? errno == ENOMEM ? NULL : strerror(errno) : (msg)))

/* Reads and returns an area of the input file.
 */
static void *getarea(int off, int size)
{
    void       *buf = NULL;

    xalloc(buf, size);
    if (fseek(infile, off, SEEK_SET) || fread(buf, size, 1, infile) != 1) {
	free(buf);
	return NULL;
    }
    return buf;
}

/* Reads in the file's ELF header, and runs several sanity checks on
 * the structure's contents. FALSE is returned if the file looks to
 * not be an ELF file at all.
 */
static int readelfhdr(void)
{
    static Elf32_Ehdr	ehdrbuf = { { 0 } };
    unsigned int	n;

    ehdr = &ehdrbuf;
    errno = 0;
    n = fread(ehdr, 1, sizeof *ehdr, infile);
    if (n <= EI_MAG3 || ehdr->e_ident[EI_MAG0] != ELFMAG0
		     || ehdr->e_ident[EI_MAG1] != ELFMAG1
		     || ehdr->e_ident[EI_MAG2] != ELFMAG2
		     || ehdr->e_ident[EI_MAG3] != ELFMAG3)
	return err("not an ELF file.");
    if (n < sizeof *ehdr) {
	err("warning: file does not contain a complete ELF header.");
	recordpiece(0, n, P_SECTION, 0, "ehdr");
    } else
	recordpiece(0, n, P_EHDR, 0, "ehdr");

    if (ehdr->e_ident[EI_CLASS] != ELFCLASS32)
	err("warning: not a 32-bit ELF file.");
    n = 1;
    *(char*)&n = 0;
    if (ehdr->e_ident[EI_DATA] != (n ? ELFDATA2MSB : ELFDATA2LSB))
	err("warning: not a %s-endian ELF file.", (n ? "big" : "little"));
    if (ehdr->e_ident[EI_VERSION] != EV_CURRENT)
	err("warning: unknown ELF header version.");
    if (ehdr->e_version != EV_CURRENT)
	err("warning: unknown ELF version.");

    if (ehdr->e_ehsize != sizeof(Elf32_Ehdr))
	err("warning: unrecognized ELF header size.");
    if (ehdr->e_phoff && ehdr->e_phentsize != sizeof(Elf32_Phdr))
	err("warning: unrecognized program header table entry size.");
    if (ehdr->e_shoff && ehdr->e_shentsize != sizeof(Elf32_Shdr))
	err("warning: unrecognized section header table entry size.");

    return TRUE;
}

/* Reads in the file's program header table and/or section header
 * table, and does some basic sanity checking. If a section header
 * table is present, the function also attempts to read the section
 * header string table.
 */
static int readhdrtables(void)
{
    int	n;

    if (ehdr->e_phoff) {
	n = ehdr->e_phnum * sizeof *phdrs;
	if ((phdrs = getarea(ehdr->e_phoff, n)))
	    recordpiece(ehdr->e_phoff, n, P_PHDRTAB, 0, "phdrs");
	else
	    err("warning: invalid program header table offset.\n");
    }

    if (ehdr->e_shoff) {
	n = ehdr->e_shnum * sizeof *shdrs;
	if ((shdrs = getarea(ehdr->e_shoff, n)))
	    recordpiece(ehdr->e_shoff, n, P_SHDRTAB, 0, "shdrs");
	else
	    err("warning: invalid section header table offset.\n");
    }

    if (!phdrs && !shdrs)
	err("warning: ELF file has no valid header tables.");

    if (shdrs && ehdr->e_shstrndx) {
	if (ehdr->e_shstrndx >= ehdr->e_shnum)
	    err("warning: invalid section header string table");
	else
	    enumsections(ehdr, shdrs,
			 getarea(shdrs[ehdr->e_shstrndx].sh_offset,
				 shdrs[ehdr->e_shstrndx].sh_size));
    }

    return TRUE;
}

/* Uses the program header table to find the file offset corresponding
 * to a memory address. -1 is returned if no such offset exists.
 */
static int addrtooffset(unsigned addr)
{
    Elf32_Phdr *phdr;
    unsigned	i;

    for (i = 0, phdr = phdrs ; i < ehdr->e_phnum ; ++i, ++phdr)
	if (phdr->p_vaddr <= addr && phdr->p_vaddr + phdr->p_memsz > addr)
	    return addr - phdr->p_vaddr + phdr->p_offset;
    return -1;
}

/* Records the contents referenced by various entries in the dynamic
 * object as separate pieces of the file. This function helps make up
 * for the absence of such information when a file contains no section
 * header table.
 */
static void getdynamicentries(Elf32_Dyn const *dynamic)
{
    Elf32_Dyn const    *dyn;
    Elf32_Sword		val[N_DT_COUNT];
    int		       *p;
    int			off, n, i;

    memset(val, 0, sizeof val);
    for (i = 0, dyn = dynamic ; dyn->d_tag != DT_NULL ; ++i, ++dyn)
	if ((n = getdyntagid(dyn->d_tag)) > 0)
	    val[n] = dyn->d_un.d_val;

#define	trysegment(addr, sz, off, type, name)				\
    (((off) = addrtooffset(addr)) >= 0					\
	&& (recordpiece((off), ((off) + (sz) > thefilesize ?		\
				     thefilesize - (off) : (sz)),	\
			(type), 0, (name)), 0))

    trysegment(val[N_DT_STRTAB], val[N_DT_STRSZ], off, P_SECTION, "~dynstr");
    trysegment(val[N_DT_SYMINFO], val[N_DT_SYMINSZ],
	       off, P_HALVES, "~syminfo");
    trysegment(val[N_DT_INIT_ARRAY], val[N_DT_INIT_ARRAYSZ],
	       off, P_WORDS, "~init_array");
    trysegment(val[N_DT_FINI_ARRAY], val[N_DT_FINI_ARRAYSZ],
	       off, P_WORDS, "~fini_array");
    trysegment(val[N_DT_PREINIT_ARRAY], val[N_DT_PREINIT_ARRAYSZ],
	       off, P_WORDS, "~preinit_array");
    if (val[N_DT_RELENT] == sizeof(Elf32_Rel))
	trysegment(val[N_DT_REL], val[N_DT_RELSZ], off, P_REL, "~rel_got");
    if (val[N_DT_RELAENT] == sizeof(Elf32_Rela))
	trysegment(val[N_DT_RELA], val[N_DT_RELASZ], off, P_RELA, "~rela_got");
    n = val[N_DT_PLTREL] == sizeof(Elf32_Rel) ? P_REL :
	val[N_DT_PLTREL] == sizeof(Elf32_Rela) ? P_RELA : 0;
    if (n)
	trysegment(val[N_DT_JMPREL], val[N_DT_PLTRELSZ], off, n,
		   n == P_REL ? "~plt_rel" : "plt_rela");

#undef trysegment

    if ((off = addrtooffset(val[N_DT_HASH])) >= 0) {
	if ((p = getarea(off, 8))) {
	    n = (p[0] + p[1] + 2) * sizeof(Elf32_Word);
	    if (off + n > thefilesize)
		n = thefilesize - off;
	    recordpiece(off, n, P_HASH, 0, "~hash");
	    off = addrtooffset(val[N_DT_SYMTAB]);
	    if (off >= 0 && val[N_DT_SYMENT] == sizeof(Elf32_Sym)) {
		n = p[1] * sizeof(Elf32_Sym);
		if (off + n > thefilesize)
		    n = thefilesize - off;
		recordpiece(off, n, P_SYMTAB, 0, "~dynsym");
	    }
	    free(p);
	}
    }
}

/* Records the contents referenced by the various entries in the
 * program header table and the section header table, each as a
 * separate piece of the file. Attempts to assign each piece an
 * appropriate name. Addresses, if present, are also recorded.
 */
static int recordsections(void)
{
    char	buf[64];
    char const *str;
    int		type;
    int		i, j;

    if (phdrs) {
	for (i = 0 ; i < ehdr->e_phnum ; ++i) {
	    switch (phdrs[i].p_type) {
	      case PT_NULL:	type = 0;				break;
	      case PT_PHDR:	type = P_PHDRTAB;  str = "~phdrs";	break;
	      case PT_DYNAMIC:	type = P_DYNAMIC;  str = "~dynamic";	break;
	      case PT_INTERP:	type = P_SECTION;  str = "~interp";	break;
	      case PT_NOTE:	type = P_NOTE;     str = "~note";	break;
	      default:		type = P_BYTES;	   str = NULL;		break;
	    }
	    if (!type)
		continue;
	    if (!str) {
		if (phdrs[i].p_flags & PF_X)
		    str = "~text";
		else if (phdrs[i].p_flags & PF_W)
		    str = "~data";
		else
		    str = "~segment";
	    }
	    j = phdrs[i].p_offset + phdrs[i].p_filesz;
	    if (j > thefilesize)
		j = thefilesize;
	    j -= phdrs[i].p_offset;
	    recordpiece(phdrs[i].p_offset, j, type, 0, str);
	    if ((phdrs[i].p_flags & PF_R) && phdrs[i].p_vaddr)
		recordaddr(phdrs[i].p_vaddr, phdrs[i].p_offset,
			   phdrs[i].p_memsz, str);
	    if (type == P_DYNAMIC)
		getdynamicentries(getarea(phdrs[i].p_offset, j));
	}
    }

    if (shdrs) {
	for (i = 0 ; i < ehdr->e_shnum ; ++i) {
	    switch (shdrs[i].sh_type) {
	      case SHT_NULL:		type = 0;			break;
	      case SHT_NOBITS:		type = 0;			break;
	      case SHT_SYMTAB:		type = P_SYMTAB;		break;
	      case SHT_DYNSYM:		type = P_SYMTAB;		break;
	      case SHT_HASH:		type = P_HASH;			break;
	      case SHT_DYNAMIC:		type = P_DYNAMIC;		break;
	      case SHT_REL:		type = P_REL;			break;
	      case SHT_RELA:		type = P_RELA;			break;
	      case SHT_NOTE:		type = P_NOTE;			break;
	      case SHT_GNU_verdef:	type = P_WORDS;			break;
	      case SHT_GNU_verneed:	type = P_WORDS;			break;
	      case SHT_GNU_versym:	type = P_HALVES;		break;
	      default:
		  if (shdrs[i].sh_entsize == 4)
		      type = P_WORDS;
		  else if (shdrs[i].sh_entsize == 2)
		      type = P_HALVES;
		  else
		      type = P_SECTION;
		  break;
	    }
	    if (!type)
		continue;
	    if ((str = getshdrtruename(i))) {
		if (type == P_SECTION && !strcmp(str, ".stab"))
		    type = P_WORDS;
		strncpy(buf, str, sizeof buf);
		buf[sizeof buf - 1] = '\0';
		for (j = 0 ; buf[j] && !isalnum(buf[j]) ; ++j) ;
		str = buf + j;
		for ( ; buf[j] ; ++j)
		    if (!isalnum(buf[j]))
			buf[j] = '_';
	    } else
		str = "~section";
	    j = shdrs[i].sh_offset + shdrs[i].sh_size;
	    if (j > thefilesize)
		j = thefilesize;
	    j -= shdrs[i].sh_offset;
	    recordpiece(shdrs[i].sh_offset, j, type, i, str);
	    if ((shdrs[i].sh_flags & SHF_ALLOC) && shdrs[i].sh_addr)
		recordaddr(shdrs[i].sh_addr, shdrs[i].sh_offset,
			   shdrs[i].sh_size, str);
	}
    }

    return TRUE;
}

/* Selects a unique name for each piece, to be used as the name of the
 * field in the C structure.
 */
static int setnames(void)
{
    int	i, j, n;

    n = 0;
    for (i = 0 ; i < piecenum ; ++i) {
	if (*pieces[i].name != '~')
	    continue;
	if (pieces[i].type == P_UNCLAIMED
		|| (pieces[i].type == P_BYTES && pieces[i].size < 16
					      && !(pieces[i].to & 3)))
	    sprintf(pieces[i].name, "pad%d", ++n);
	else
	    memmove(pieces[i].name, pieces[i].name + 1,
				    sizeof pieces->name - 1);
    }

    for (i = 0 ; i < piecenum - 1 ; ++i) {
	n = 1;
	for (j = i + 1 ; j < piecenum ; ++j)
	    if (!strcmp(pieces[i].name, pieces[j].name))
		sprintf(pieces[j].name, "%s%d", pieces[i].name, ++n);
	if (n > 1)
	    strcat(pieces[i].name, "1");
    }

    setaddrnames();

    return TRUE;
}

/* main().
 */
int main(int argc, char *argv[])
{
    void       *buf;
    struct stat	st;
    int		i;

    thisprog = argv[0];
    structname = "elf";
    varname = "foo";
    outfile = stdout;

    if (argc != 2) {
	err("Usage: elftoc ELFFILENAME");
	return argc < 2;
    }
    thefilename = argv[1];

    if (stat(thefilename, &st)) {
	errnomsg("can't stat file");
	return 1;
    }
    thefilesize = st.st_size;

    if (!(infile = fopen(thefilename, "r"))
		|| !readelfhdr()
		|| !readhdrtables())
	return 1;

    recordpiece(0, thefilesize, P_UNCLAIMED, 0, "~unused");
    recordsections();
    arrangepieces();
    verifypiecetypes();

    setnames();
    beginoutpieces();
    for (i = 0 ; i < piecenum ; ++i) {
	buf = getarea(pieces[i].from, pieces[i].size);
	outpiece(pieces + i, buf);
	free(buf);
    }
    endoutpieces();

    return 0;
}
